keps/sig-apps/2307-job-tracking-wihout-lingering-pods/README.md

```diff
@@ -0,0 +1,599 @@
+intermediate state is necessary. Before removing the Pod finalizers,
+the controller adds finished Pods to a list in the Job status. After removing
+the Pod finalizers, the controller clears the list and updates the counters.
+
+### Risks and Mitigations
+
+1. The new implementation produces new API calls: one per Pod to remove
+   finalizers and one for each Job sync, to track the intermediate state.
+   However, the ability to handle higher number of Pods and Jobs at a given
+   point justifies the added cost of the new API calls.
+2. Changes in the status not produced by the Job controller in
+   kube-controller-manager could affect the Job tracking. Cluster administrators
+   should make sure to protect the Job status endpoint via RBAC.
+
+## Design Details
+
+### API changes
+
+The Job status gets a new struct to hold the uncounted Pods before they are
+added to the counters.
+
+```golang
+type JobStatus struct {
+    Succeeded int32
+    Failed    int32
+    ...
+
+    // UncountedPodUIDs holds UIDs of Pods that have finished but haven't
+    // been accounted in the counters.
+    UncountedPodUIDs UncountedPodUIDs
+}
+
+// UncountedPodUIDs holds UIDs of Pods that have finished but haven't been
+// accounted in Job status counters.
+type UncountedPodUIDs struct {
+    // Enabled indicates whether Job tracking uses this struct.
+    Enabled   bool
+    // Succeeded holds UIDs of succeeded Pods.
+    Succeeded []string
+    // Succeeded holds UIDs of failed Pods.
+    Failed    []string
+}
+```
+
+### Algorithm
+
+The following algorithm updates the status counters without relying on finished
+Pods to be present indefinitely. The algorithm assumes that the Job controller
+could be stopped at any point and executed again from the first step without
+losing information. Generally, all the steps happen in a single Job sync
+cycle.
+
+1. The Job controller creates Pods for a Job with the finalizer
+   `batch.kubernetes.io/job-completion`
+2. The Job controller adds Pod UIDs to the `.status.uncountedPodUIDs.succeeded`
+   and `.status.uncountedPodUIDs.failed` lists if the Pod:
+    - has the `batch.kubernetes.io/job-completion` finalizer, and
+    - the Pod is on Succeeded or Failed phase, respectively.
+    The controller sends a status update.
+3. The Job controller removes the `batch.kubernetes.io/job-completion` finalizer
+   from all Pods on Succeeded or Failed phase that were added to the lists in
+   `.status.uncountedPodUIDs` in the previous step. The controller keeps note
+   of the Pod updates that succedeed.
+4. The Job controller counts the Pods in the `.status.uncountedPodUIDs` lists
+   that:
+   - had no finalizer at the beginning of the sync, or
+   - had the patch in step 3 succeed, or
+   - were removed from the system.
+   The counts increment the `.status.failed` and `.status.succeeded` and clears
+    counted Pods from `.status.uncountedPodsUIDs` lists. The controller sends a
+    status update.
```

Comment:
Talk about if there are changes to the logic where the controller decides whether to start new pods.
Does it look at the current `.status.succeeded` and compare that to `.spec.completions`?  Or does it look at  `.status.succeeded + len( status.uncountedPodUIDs.succeeded`?  The list could be non-empty due to the throttling mentioned below.  Or does it block starting new pods if there are any uncounted pods?

Reply:
This is a good point, "uncounted" isn't quite right, it's "terminatedAndUncounted". Yes, that logic needs to look at 3 sources of information, a) the counts, b) the existing pods which are not in c) the terminatedAndUncounted list.

Reply:
changed